File: replay/backend/app/api/routes/paper.py
"""Paper-trading : ticket, compte, positions, ordres simulés (toujours dry_run, aucun broker)."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from app.api.dependencies import Container, get_container
from app.schemas.paper import BreakevenRequest, ModifyRequest, PartialRequest, PlaceRequest

router = APIRouter(prefix="/paper", tags=["paper"])


@router.get("/ticket")
async def ticket(symbol: str = Query("NQ"), c: Container = Depends(get_container)) -> dict:
    return c.paper.ticket(symbol)


@router.get("/account")
async def account(c: Container = Depends(get_container)) -> dict:
    return {"ok": True, "account": c.paper.account()}


@router.get("/positions")
async def positions(c: Container = Depends(get_container)) -> dict:
    return {"positions": c.paper.positions(), "orders": []}


@router.post("/place")
async def place(payload: PlaceRequest, c: Container = Depends(get_container)) -> dict:
    return await c.paper.place(
        payload.symbol,
        payload.side,
        payload.size,
        payload.sl_price,
        payload.tp_price,
        payload.entry_ref,
        payload.entry_tag,
    )


@router.post("/flatten")
async def flatten(c: Container = Depends(get_container)) -> dict:
    return await c.paper.flatten()


@router.post("/breakeven")
async def breakeven(payload: BreakevenRequest | None = None, c: Container = Depends(get_container)) -> dict:
    return await c.paper.breakeven((payload or BreakevenRequest()).offset_ticks)


@router.post("/partial-close")
async def partial_close(payload: PartialRequest, c: Container = Depends(get_container)) -> dict:
    return await c.paper.partial_close(payload.size)


@router.post("/modify")
async def modify(payload: ModifyRequest, c: Container = Depends(get_container)) -> dict:
    return await c.paper.modify(payload.which, payload.price)


@router.post("/reset")
async def reset(c: Container = Depends(get_container)) -> dict:
    return await c.paper.reset()

File: replay/frontend/lib/api/paper.ts
// Paper-trading : mutations explicites, jamais rejouées automatiquement en cas de timeout.
import { api, qs } from './client';
import type { Account, Position, TicketSpec } from '$replay/domain/market/types';

export interface PlacePayload {
	symbol: string; side: 'buy' | 'sell'; size: number; sl_price: number | null; tp_price: number | null;
	entry_ref: number | null; risk_usd: number; entry_tag?: string | null;
}
export interface PlaceResult { ok: boolean; fill: number; sl_points: number; tp_points: number; regime: string | null; setup: string | null; dry_run: boolean; }

export class PaperApi {
	static ticket(symbol: string) { return api.get<TicketSpec>(`/paper/ticket${qs({ symbol })}`); }
	static account() { return api.get<{ account: Account }>('/paper/account').then((r) => r.account); }
	static positions() { return api.get<{ positions: Position[] }>('/paper/positions').then((r) => r.positions); }
	static place(payload: PlacePayload) { return api.post<PlaceResult>('/paper/place', payload); }
	static flatten() { return api.post<{ ok: boolean; closed: boolean; pnl: number }>('/paper/flatten'); }
	static breakeven(offset_ticks = 0) { return api.post<{ ok: boolean; entry: number; sl: number }>('/paper/breakeven', { offset_ticks }); }
	static partialClose(size: number) { return api.post<{ ok: boolean; closed: number; remaining: number; pnl: number }>('/paper/partial-close', { size }); }
	static modify(which: 'sl' | 'tp', price: number) { return api.post<{ ok: boolean; price: number }>('/paper/modify', { which, price }); }
	static reset() { return api.post<{ ok: boolean; account: Account }>('/paper/reset'); }
}
